File: 897.IncreasingOrderSearchTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var increasingBST = function(root) {
    let result = null;
    const stack = [];
    let current = root;
    let pointer;
    while(true) {
        if (current) {
            stack.push(current);
            current = current.left;
        } else if (stack.length) {
            const node = stack.pop();
            if (!result) {
                result = pointer = new TreeNode(node.val);  
            } else {
                pointer.left = new TreeNode(node.val);  
                pointer = pointer.left;
            }
            current = node.right;
        } else {
            break;
        }
    }
    return result;
};

File: Golang/Tree/897.Increasing-order-search-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func increasingBST(root *TreeNode) *TreeNode {
	var newRoot, prev *TreeNode
	current := root
	stack := []*TreeNode{}
	for {
		if current != nil {
			stack = append(stack, current)
			current = current.Left
		} else if len(stack) > 0 {
			node := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			if newRoot == nil {
				newRoot = &TreeNode{
					Val: node.Val,
				}
				prev = newRoot
			} else {
				prev.Right = &TreeNode{
					Val: node.Val,
				}
				prev = prev.Right
			}
			current = node.Right
		} else {
			break
		}
	}
	return newRoot
}
